### PR TITLE
Video 37: "In Search of Code Purity"

### PR DESCRIPTION
**Pull request by 0atman**
2024-01-16T16:24:00,159033644+00:00

### CODE REVIEW

File: scripts/37-functional-rust.md [hunk 1]
```diff
@@ -0,0 +1,1189 @@
...
+```rust
+// #[cfg(test)] omitted because I want linting
+mod tests {
+
+#[allow(clippy::wildcard_imports)]
+use rstest::*;
+
+#[fixture]
+fn setup() {
+	color_eyre::install().expect("color_eyre installed");
```
Comment (laundmo) [inline scripts/37-functional-rust.md:539]:
```suggestion
	color_eyre::install().unwrap_or_else(|report| std::process::exit(<report as std::process::Termination>.report()));
```

Probably the better idea is to just put "could not install eyre" in the message, but i wanted to know how to do this:
it uses the Termination trait to process the report in a status code and exits the process with it. this is what Rust does in the background when you return a Result<(), Report> from `main` afaict

File: scripts/37-functional-rust.md [hunk 2]
```diff
@@ -0,0 +1,1189 @@
...
+
+So, when debugging, the only way to get information out of a const function is by its return type or halting compilation with a panic! or a failed assertion, like in the first code block here.
+
+Because there's no runtime difference with making a function const, most of the standard library's functions that can safely be called from a const context are already marked const and ready to use.
+
+So printing isn't possible, but if you take my advice, you'll debug with the compiler, not printing to the screen, and I love Rust's compiler driven development workflow!
+
+Note that only simple error messages available in const assertions, you can't use string formatting to interpolate values.
+
+But, in the second code block, the `static_assertions` crate provides const versions of assertions that panic with an error message with variables and other useful context.
```
Comment (laundmo) [inline scripts/37-functional-rust.md:995]:
```suggestion
But, in the second code block, the `const_panic` crate provides const versions of assertions that panic with an error message with variables and other useful context.
```

**Review by laundmo — CHANGES_REQUESTED** [review]

File: scripts/37-functional-rust.md (hunk 1 above)
Comment (0atman) [reply scripts/37-functional-rust.md:539]: oh nice! I might just allow an exception to `unwrap_used` here and call it a day with an `.unwrap()` (for brevity)

Comment (0atman) [reply scripts/37-functional-rust.md:539]: Also, @laundmo, are you in the "errors go in expect() messages" camp, not "my expectations go in expect() messages" camp?!

I'm astonished :wink: https://stackoverflow.com/questions/66362625/why-is-rusts-expect-called-expect

Comment (laundmo) [reply scripts/37-functional-rust.md]: i am in the "the error i expect to trigger this" camp

**Pull request merged**